src/Security/TOTPManager.php

```diff
@@ -0,0 +1,495 @@
+        ])->current()['2fa'] ?? null;
+        if ($tfa === null) {
+            return [];
+        }
+        $tfa = json_decode($tfa, true, 512, JSON_THROW_ON_ERROR);
+        $tfa['backup_codes'] = $code_hashes;
+        $DB->update('glpi_users', [
+            '2fa' => json_encode($tfa, JSON_THROW_ON_ERROR),
+        ], [
+            'id' => $users_id
+        ]);
+        return $random_codes;
+    }
+
+    /**
+     * Verify a backup code for a user
+     * @param string $code Backup code
+     * @param int $users_id User ID
+     * @param bool $consume_code If true, the backup code will be consumed and will not be usable again
+     * @return bool True if the code is valid, false otherwise
+     * @throws \JsonException
+     */
+    public function verifyBackupCodeForUser(string $code, int $users_id, bool $consume_code = true): bool
+    {
+        global $DB;
+
+        $tfa = $DB->request([
+            'SELECT' => ['2fa'],
+            'FROM' => 'glpi_users',
+            'WHERE' => [
+                'id' => $users_id
+            ],
+            'LIMIT' => 1
+        ])->current()['2fa'] ?? null;
+        if ($tfa === null) {
+            return false;
+        }
+        $tfa = json_decode($tfa, true, 512, JSON_THROW_ON_ERROR);
+        if (!isset($tfa['backup_codes'])) {
+            return false;
+        }
+        foreach ($tfa['backup_codes'] as $i => $hash) {
+            if (password_verify($code, $hash)) {
+                if ($consume_code) {
+                    unset($tfa['backup_codes'][$i]);
+                    $DB->update('glpi_users', [
+                        '2fa' => json_encode($tfa, JSON_THROW_ON_ERROR),
+                    ], [
+                        'id' => $users_id
+                    ]);
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Get the 2FA enforcement level for a user
+     * @param int $users_id ID of the user
+     * @return int One of the ENFORCEMENT_* constants
+     * @phpstan-return self::ENFORCEMENT_*
+     * @throws \Exception
+     */
+    public function get2FAEnforcement(int $users_id): int
+    {
+        global $CFG_GLPI, $DB;
+
+        $user_optional = $DB->request([
+            'SELECT' => ['2fa_unenforced'],
+            'FROM' => 'glpi_users',
+            'WHERE' => [
+                'id' => $users_id
+            ],
+            'LIMIT' => 1
+        ])->current()['2fa_unenforced'] ?? 0;
+
+        if ($user_optional) {
+            return self::ENFORCEMENT_OPTIONAL;
+        }
+
+        $enforced = false;
+        if ((int) $CFG_GLPI['2fa_enforced'] === 1) {
+            $enforced = true;
+        } else {
+            // Check entity configuration
+            $entities = PermissionManager::getInstance()->getAllEntities($users_id);
+            foreach ($entities as $entity) {
+                if ((int) \Entity::getUsedConfig('2fa_enforcement_strategy', $entity, '', true) === 1) {
```

Comment:
the `true` value for default make the `get2FAEnforcement` method to return enforced=true when no enforcement has been set at all.

And so, i didn't setup the enforcement while testing the feature on an instance where the pr hasn't been tested at all, and it asked to register the 2fa application